Bluetooth: Audio - media control kconfig refactor

asbjornsabo opened the pull request:
This PR refactors Bluetooth audio media control Kconfig files. (There is also a small related clean up.)

The overall purpose is to get a clear high-level media control configuration, to separate modules more clearly and to remove Kconfig dependencies
- The media player gets a separate Kconfig
- There is new top-level media control Kconfig that outlines the main options and enables media control service and client if needed.
- The media proxy implementation is better guarded by configs, so that only used code is compiled

asbjornsabo committed:
Bluetooth: Audio: Media control Kconfig

Add Kconfig file for the media proxy, for high-level media control
configuration.

The media proxy did not have a Kconfig file, it was relying on the
Kconfigs for GATT media control client and server.

The purpose is to unify the media control configurations, to make the
different media control functionalities available separately, and to
separate the high-level media control from the bluetooth media control
server and client.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Separate Kconfig for MPL

Separate out the Kconfig for the media player.
Also removes some unused MCS configs.

The media player has been using MCS configs, while it should be the
other way around.

Also, this will make it easier to move the media player elsewhere, as
has been discussed.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio - remove media controller dependency

The media controller internal header file does not need to depend upon
the media proxy header.

Requires one piece of spec info here, but the media player already
needs to adhere to the media control service spec, so situation no
worse than before.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: Use new media control configs in media proxy

Update the media proxy to use the new Kconfigs to include/exclude the
various parts of the implementation. This replaces existing guarding
using other Kconfig values.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio: MCS config file cleanup

- Rename MCS configs that are only used by MCC to "MCC", add some more
  documentation.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

Thalley commented on subsys/bluetooth/audio/CMakeLists.txt:
```diff
@@ -38,9 +38,9 @@ zephyr_library_sources_ifdef(CONFIG_BT_MCC mcc.c)
+zephyr_library_sources_ifdef(CONFIG_BT_MPL mpl.c)
 
-zephyr_library_sources_ifdef(CONFIG_BT_MCS media_proxy.c)
+zephyr_library_sources_ifdef(CONFIG_MCTL media_proxy.c)
```
I am quite unsure about the mapping between media proxy and media control. While they are very similar, it just reads weird that control enables a proxy. Suggest to unify the names.

asbjornsabo replied:
I have been thinking along the same lines.  (Does that mean that we are great minds? :-) )
I have been thinking that the media_proxy could be renamed to media_control.  (That would be a separate PR, I think.)
Opinions?

Thalley replied:
I actually like "media control" more than "media proxy" :)

Thalley replied:
Please create a GH issue to rename it if we want to do that, or rename the Kconfig to match "proxy"

asbjornsabo replied:
https://github.com/zephyrproject-rtos/zephyr/issues/43107

Thalley commented on subsys/bluetooth/audio/Kconfig.mcs:
```diff
@@ -51,6 +37,38 @@ config BT_MCC
+	default 20
+	range 1 255
+	help
+	  Sets the maximum number of bytes (including the null termination) of
+	  the name of the media player.
```
Perhaps it makes sense to add what happens if the client receives a name larger than this value (truncation or discarding).

Thalley replied:
Also applies the configs below

Thalley commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,84 @@
+	  This option enables support for remote control of local media
+	   players
+
+config MCTL_LOCAL_PLAYER_REMOTE_CONTROL_BT
```
If we keep this for BT-only we can omit the BT postfix (but will need a BT_ prefix). 

If we want to use this for non-BT, then I strongly suggest that e.g. MCS enables this, instead of the other way around.

asbjornsabo replied:
After thinking more, I partly changed my mind, and found that we can at least keep the possibility to allow other protocols.  Therefore the _BT suffix - there is now room to add other protocols without disturbing what is there.

As for which module enables which, it is not obvious to me that the protocol (BT in this case) should be the "center of the world".  But that is a discussion I think we can take separately.  The important thing now is that the configs for the media_proxy, the player and the MCS/MCC are separated, with the corresponding code changes.  Switching the direction of dependencies is then easily done just by modifying Kconfig files.

Thalley replied:
Agreed, this is outside the scope of this PR. However, in that case I'd still remove the _BT suffix, as that is making it more confusing than not at this point. 

Will you create a GH issue to track the potential refactoring of that?

asbjornsabo replied:
Reopened https://github.com/zephyrproject-rtos/zephyr/issues/42523 and added to the description.

Removing the _BT suffix defeats the purpose of the splitting, as it will tie the (generic) remote control option to the Bluetooth protocol.  The option would be to remove the non_BT general option, and put it back when needed, but that hides how the structure is intended to be.
Tried to make less confusing by adding to the help texts.

Thalley replied:
> Removing the _BT suffix defeats the purpose of the splitting,

I disagree. If the media control/proxy is a separate feature from BT, then it should be split regardless. 
However, my opinion still stands that if the media control/proxy is separate from BT, then it should be responsible for enabling/disabling any BT related features, but rather have the BT related features `depend on` or `select` this instead, as we do for `BT_SETTINGS`. 

Having BT specific Kconfigs in a non-BT feature is not the right way to go about this in my opinion. 

Perhaps we should get a 3rd or 4th set on eyes on this as well.

asbjornsabo replied:
I wonder if we may be talking past each other.
Is it the use of "select" you are opposed to?  (That is something to be discussed, but as I said that can easily be fixed later.)
Or is it the _BT suffix for the Bluetooth remote controls?

asbjornsabo replied:
Changed "select" to "depends on"

Thalley commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,84 @@
+	  This options enables support for accessing objects on a remote media
+	  player
+
+config BT_DEBUG_MEDIA_PROXY
```
Naming scheme does not match the other Kconfigs options

Thalley commented on subsys/bluetooth/audio/Kconfig.mpl:
```diff
@@ -0,0 +1,115 @@
+	  Use this option to set the name of the media player.
+
+config BT_MPL_MEDIA_PLAYER_NAME_MAX
+	# TODO: Remove - having the name (above) should give the length
```
Could we remove it in this PR? Seems pretty trivial

Thalley replied:
Ditto for the ones below

asbjornsabo replied:
I have reconsidered (and removed the TODOs), as i see that these are actually needed.
For the GMCS (which we have), the name can change runtime, when the player changes.  
If that happens, the URL may also change.

(The reason I kept it out of this PR was that that would be a code change, while this PR deals with config changes.)

asbjornsabo commented:
Latest push:
Kconfig.mctl: Changed "select" to "depends on"
Kconfig.mpl/mcs are sourced directly from the main Kconfig file

asbjornsabo committed:
Bluetooth: Audio: Fix initialization of not present module

Do not initialize the media control service if it is not configured.

This is a temporary fix, there is an issue to fix this properly:
https://github.com/zephyrproject-rtos/zephyr/issues/42965

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
And one more push, to fix some dependency issues that cropped up (planned to be fixed) that broke twister and to add a TODO.

Thalley commented on subsys/bluetooth/audio/Kconfig.mcs:
```diff
@@ -51,6 +37,43 @@ config BT_MCC
+	range 1 255
+	help
+	  Sets the maximum number of bytes (including the null termination) of
+	  the name of the media player.  If the name is longer, the media
```
```suggestion
	  the name of the media player. If the name is longer, the media
```
Here and also in the following configs

asbjornsabo replied:
Fixed

Thalley commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,83 @@
+	  Currently, the only option for remote player control is via Bluetooth.
+	  More options may be added later.
+
+config MCTL_REMOTE_PLAYER_CONTROL_BT
+	bool "Support for control of remote media player via Bluetooth"
+	default y
+	depends on MCTL_REMOTE_PLAYER_CONTROL
+	depends on BT_MCC
+	help
+	  This options enables support for control of a remote media player,
+	  using GATT and the Media Control Client
```
Ditto, BT_MCC should depend on (possibly, since the MCC actually have it's own API it doesn't really depend on this) `MCTL_REMOTE_PLAYER_CONTROL_BT`, and remove the `_BT` suffix.

Thalley commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,83 @@
+	  This options enables support for control of a remote media player,
+	  using GATT and the Media Control Client
+
+config MCTL_REMOTE_PLAYER_CONTROL_BT_OBJECTS
+	bool "Support for accessing objects on remote player"
+	depends on MCTL_REMOTE_PLAYER_CONTROL_BT
+	default y
+	depends on BT_MCC_OTS
+	help
+	  This options enables support for accessing objects on a remote media
+	  player
```
It feels like there is still some miscommunication. My take from the meeting was that generic (i.e. non-BT specific) features/subsystems should have have any BT specific dependencies nor config options. 

Furthermore this config option feels too generic: What is an "object" in a media controller?

asbjornsabo replied:
Clarified the objects.

Thalley commented on tests/bluetooth/shell/audio.conf:
```diff
@@ -64,24 +64,37 @@ CONFIG_BT_CSIS=y
+CONFIG_MCTL_REMOTE_PLAYER_CONTROL_BT_OBJECTS=y
+
+CONFIG_BT_MPL=y
+CONFIG_BT_MPL_MEDIA_PLAYER_NAME_MAX=30
+CONFIG_BT_MPL_MEDIA_PLAYER_NAME="My media player"
+CONFIG_BT_MPL_ICON_URL_MAX=27
+CONFIG_BT_MPL_ICON_URL="http://www.somecompany.com"
+CONFIG_BT_MPL_TRACK_TITLE_MAX=40
+CONFIG_BT_MPL_SEGMENT_NAME_MAX=35
```
Suggest to use default values unless necessary.

asbjornsabo committed:
Bluetooth: Audio - MCS kconfig - fix spacing in comments

Change double-space after period to single space.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,83 @@
+	  This options enables support for accessing objects on a remote media
+	  player
+
+config BT_DEBUG_MCTL
```
Agree - that one had slipped through

asbjornsabo committed:
Bluetooth: Audio: Media - change default kconfig values

Change the default URL length value to something that can contain the
default URL string.
Increaste the default track title length.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio - media - do not use non-default values in settings

Remove some project-specific settings for media, rely on defaults
instead.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.mcs:
```diff
@@ -51,6 +37,43 @@ config BT_MCC
+
+config BT_MCC_ICON_URL_MAX
+	int "Max length of media player icon URL"
+	default 30
```
Changed

asbjornsabo committed:
Bluetooth: Audio - media - guard instance members

Guard media proxy instance members.
As a consequence, change guarding for functions using remote player
member.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo committed:
Bluetooth: Audio - media - add more guarding

Add include guards for local player local control

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
Pushed - pure rebase on main.

Thalley commented:
> Pushed - pure rebase on main.

Please rebase again to solve conflicts :)

asbjornsabo commented:
Pushed again - rebased on main and resolved the conflicting changes from the OTS restructure that had been merged.

Thalley commented on subsys/bluetooth/audio/mpl_internal.h:
```diff
@@ -10,54 +10,53 @@
  * are used when changing to/from zero
  */
-#define MPL_SEEKING_SPEED_FACTOR_STEP  BT_MCS_SEEKING_SPEED_FACTOR_MIN
+#define MPL_SEEKING_SPEED_FACTOR_STEP  4
```
Should this actually have been a Kconfig?

asbjornsabo replied:
This is left unspecified (on purpose) in the Media Control Service specification.  And it is likely to be very implementation specific.  So I see no value in making it configurable for this sample player.

Thalley replied:
Everything related to the MPL is implementation specific, and yet we have plenty of Kconfigs for it already (including the player name, etc.), so I do not follow your argument. 

However, given that the MPL will be moved to `samples`, then Kconfigs are not necessary in any case.

asbjornsabo replied:
I think the argument is more or less like this: The MCS spec says that there shall be a media player name (in the service).  So it makes sense to have a name for the player.  But for the seeking speed behaviour, the MCS deliberately leaves almost everything up to the implementation.

Thalley replied:
The media player name, and the media player seeking speed factor are both 2 values that reside in the media player and are implementation specific. I don't see why they should be treated different in any way. Whether they are controlled via Kconfigs or hardcoded in header files doesn't really matter once this is moved to a sample application.

Thalley commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,66 @@
+	help
+	  This options enables support for control of a remote media player.
+
+config MCTL_REMOTE_PLAYER_CONTROL_OBJECTS
```
This one is interesting. Have you had any thoughts on how this will become something generic (i.e. non-BT MCS)?

asbjornsabo replied:
That one is interesting, indeed.  This is related to the other things that must be cleared up regarding objects.
But all the objects are stored in the player, so one option could be to add a (local) API to the player for object access.

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,66 @@
+
+config MCTL_LOCAL_PLAYER_CONTROL
+	bool "Support for control of local media player"
+	# TODO: Move MPL out of BT susbystem
+	depends on BT_MPL
```
Actually, it does not have to - there does not have to be a local player.
So probably MPL should depend on MCTL instead.

asbjornsabo replied:
I was thinking of MCTL in general.  This is for the "LOCAL_PLAYER_CONTROL", and that does indeed depend upon the MPL.  There can be no control of a local player if there is no local player.
But it is a "depends" and not a "select", as per the previous discussion.

There should also be a dependency the other way - MPL is of no use if there is no interface for it.

asbjornsabo replied:
And one more thing: If we move this to a sample application, it will probably be moved along with the media_proxy, to the same subsystem.  So at least this dependency will be within that subsystem.

Thalley replied:
> And one more thing: If we move this to a sample application, it will probably be moved along with the media_proxy, to the same subsystem. So at least this dependency will be within that subsystem.

A subsystem feature (e.g. MCTL) cannot depend on a sample application (e.g. the MPL). Since the MPL is not in the stack, the MCTL should not have a dependency towards it. 

It could be argued that the BT subsystem does not work if there's no application that calls bt_enable, but that doesn't mean that it cannot be enabled. I think it's basically the same case here.

asbjornsabo replied:
I would say the case is more like the example you mentioned, where Bluetooth settings depends upon SETTINGS.
There is a difference between "not being used" and "not being able to run".

asbjornsabo replied:
Oh, well, that was not precise.  It will run, it just won't be able to do anything.

asbjornsabo replied:
I see your point.  It will be able to build, and it is not strictly depending on the MPL - if there is no MPL, no player will be registered, and the application will just get error codes in return.

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,66 @@
+	help
+	  This option enables support for control of a local media player
+
+config MCTL_LOCAL_PLAYER_LOCAL_CONTROL
+	bool "Support for local control of local media player"
+	depends on MCTL_LOCAL_PLAYER_CONTROL
+	help
+	  This option enables support for local application control of local
+	   media players
+
+config MCTL_LOCAL_PLAYER_REMOTE_CONTROL
```
That makes sense, I agree.
Propose to make the control of remote player the default, as Zephyr is more likely to be the media control client.

And propose to make the local control of the local player the default, if there is a local player I believe one will want to control it locally.

asbjornsabo replied:
Actually, MCTL_REMOTE_PLAYER_CONTROL is already default "y"

asbjornsabo commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,66 @@
+
+if MCTL
+
+config MCTL_LOCAL_PLAYER_CONTROL
```
I prefer to keep the infix
- one does not have to look at the other alternative
- it keeps the naming scheme the same for the two alternatives
- if the infix is removed, the meaning would basically be the same as the MCTL a level above - it is all about media player control after all.

asbjornsabo committed:
Bluetooth: Audio - media: Make local control of local player default

Make local control of local player the default if control of local
player is configured.

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
Rebased on main - only one small adaption due to the addition of a .connected callback in main.
Added one more commit.

asbjornsabo commented:
Rebased - fixed merge conflict in Kconfig due to the BASS PR (https://github.com/zephyrproject-rtos/zephyr/pull/40671) having been merged.

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.mctl:
```diff
@@ -0,0 +1,67 @@
+	help
+	  This option enables support for control of a local media player
+
+config MCTL_LOCAL_PLAYER_LOCAL_CONTROL
+	bool "Support for local control of local media player"
+	default y
+	depends on MCTL_LOCAL_PLAYER_CONTROL
+	help
+	  This option enables support for local application control of local
+	   media players
```
Does it have anything with Bluetooth?

asbjornsabo replied:
It has to do with Bluetooth in the sense that Bluetooth is used for remote control of the local player, or for control of a remote player.
It is likely that this module may be moved out of the Bluetooth subsystem.  But that is not part of this PR.

asbjornsabo commented:
Removed dependency on BT_MPL

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.mpl:
```diff
@@ -0,0 +1,116 @@
+
+endif # BT_MPL_OBJECTS
+
+
```
redundant empty line

asbjornsabo replied:
Removed

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.mpl:
```diff
@@ -0,0 +1,116 @@
+
+
+config BT_DEBUG_MPL
+	bool "Media Control Service debug"
```
Media Player Service debug

MariuszSkamra reviewed (commented):
Two minor issues found, otherwise LGTM

asbjornsabo committed:
Bluetooth: Audio: Bugfix - add cast to conn pointer

Cast conn pointer to void in debug statement, to avoid compilation
error.

(Explanation given: Trying to log an opaque pointer causes error.)

Signed-off-by: Asbjørn Sæbø <asbjorn.sabo@nordicsemi.no>

asbjornsabo commented:
Push:
- Rebase on main
- Add additional first commit fixing issue from logging opaque pointer that came in from rebase
- Fix two comments from Mariusz
